grpc/src/main/java/io/stargate/grpc/service/streaming/MessageStreamObserver.java

```diff
@@ -0,0 +1,97 @@
+package io.stargate.grpc.service.streaming;
+
+import com.google.protobuf.GeneratedMessageV3;
+import io.grpc.stub.StreamObserver;
+import io.stargate.grpc.service.ExceptionHandler;
+import io.stargate.grpc.service.MessageHandler;
+import io.stargate.grpc.service.SuccessHandler;
+import io.stargate.proto.QueryOuterClass;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+
+/**
+ * Implements the {@link StreamObserver} interface. It is able to process Query and Batch (both
+ * extends {@link GeneratedMessageV3}).
+ *
+ * @param <MessageT> - type of the Message to process. We support {@link
+ *     io.stargate.proto.QueryOuterClass.Query} and {@link io.stargate.proto.QueryOuterClass.Batch}
+ */
+public class MessageStreamObserver<MessageT extends GeneratedMessageV3>
+    implements StreamObserver<MessageT>, SuccessHandler {
+
+  private final AtomicLong inFlight = new AtomicLong(0);
+  private final AtomicBoolean clientSignalComplete = new AtomicBoolean(false);
+  private final ExceptionHandler exceptionHandler;
+  StreamingHandlerFactory<MessageT> streamingHandlerFactory;
+  private final StreamObserver<QueryOuterClass.StreamingResponse> responseObserver;
+
+  public MessageStreamObserver(
+      StreamObserver<QueryOuterClass.StreamingResponse> responseObserver,
+      Function<SuccessHandler, ExceptionHandler> exceptionHandlerProducer,
+      StreamingHandlerFactory<MessageT> streamingHandlerFactory) {
+    this.responseObserver = responseObserver;
+    this.streamingHandlerFactory = streamingHandlerFactory;
+    this.exceptionHandler = exceptionHandlerProducer.apply(this);
+  }
+
+  /**
+   * Calls {@code StreamObserver#onNext} for a response. Next, it decrements the {@code inFlight}
+   * and checks if the {@code this#onCompleted()} was called. If it was (and there were no errors),
+   * it calls {@code onCompleted()} on the {@code this#responseObserver}.
+   *
+   * @param response
+   */
+  @Override
+  public void handleResponse(QueryOuterClass.StreamingResponse response) {
+    try {
+      responseObserver.onNext(response);
+    } finally {
+      if (inFlight.decrementAndGet() == 0 && clientSignalComplete.get()) {
+        responseObserver.onCompleted();
+      }
+    }
+  }
+
+  /**
+   * It is creating new handler for each request and invokes the {@link MessageHandler#handle()}
+   * method.
+   *
+   * @param value Batch or Query
+   */
+  @Override
+  public void onNext(MessageT value) {
+    inFlight.incrementAndGet();
+    streamingHandlerFactory.create(value, this, exceptionHandler).handle();
```

Comment:
why are we creating the handler object on every request? Is this really designed correctly, that we have a handler per query? how does this influence the GC? shouldn't the query be parameter of the handle method? Or is there a specific reason for this?

Reply:
this is the current design where the handler is per request. We would need @mpenick input on why it was designed this way. We may consider investigating re-design but this is definitely out of the scope of this ticket.

Reply:
Can we at least open an issue as a starting point?

Reply:
done